keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,663 @@
+-->
+
+## Motivation
+
+Pod affinity/anti-affinity API allows using this feature across namespaces using 
+a static list, which works if the user knowns the namespaces names in advance.
+
+However, there are cases where the namespaces are not known beforehand, for which there
+is no way to use pod affinity/anti-affinity. Allowing users to specify the set of namespaces
+using a namespace selector addresses this problem.
+
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Allow users to dynamically select the set of namespaces considered when using pod
+  affinity/anti-affinity
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+<!--
+### Non-Goals
+
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I am running a SaaS service, the workload for each customer is placed in a separate
+namespace. The workloads requires 1:1 pod to node placement. I want to use pod 
+anti-affinity across all customers namespaces to achieve that.
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+The main risk is performance, using namespace selector will make it easier for 
+users to specify affinity constraints across a large number of namespaces.
+
+The initial implementation of pod affinity/anti-affinity suffered from 
+performance challenges, however over the releases 1.17 - 1.20 we significantly
+improved that. 
+
+We currently have integration test and clusterloader benchmarks that evaluate the extreme
+cases of all pods having affinity/anti-affinity constraints to/against each other.
```

Comment:
with this proposal, the worst-case would now involve a namespace query and iterating over all the pods namespace-by-namespace

Reply:
We currently iterate over all the pods on each node, and for each pod we test whether or not it is in an eligible namespace. 

However, there is indeed a performance concern when evaluating existing pods anti-affinity constraints: while for the incoming pod we can pre-calculate the set of namespaces, it might be expensive to do that for each existing pod when evaluating their anti-affinity constraints, pre-calculating and storing caching those for scheduled pods is not an option because the number of namespaces could be large, and because the labels of those namespaces could change.

Ideally the implementation shouldn't influence the API, but what are your thoughts on changing the API to a regexp field? it should be much faster to process than looking up namespaces via label selector.

Reply:
So this doesn't change how we iterate.